// @aws-cdk/aws-apigatewayv2-integrations/lib/http/http-proxy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpProxyIntegration = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_apigatewayv2_1 = require("@aws-cdk/aws-apigatewayv2");
/**
 * (experimental) The HTTP Proxy integration resource for HTTP API.
 *
 * @experimental
 */
class HttpProxyIntegration {
    /**
     * @experimental
     */
    constructor(props) {
        this.props = props;
        jsiiDeprecationWarnings._aws_cdk_aws_apigatewayv2_integrations_HttpProxyIntegrationProps(props);
    }
    /**
     * (experimental) (experimental) Bind this integration to the route.
     *
     * @experimental
     */
    bind(_) {
        var _b;
        return {
            method: (_b = this.props.method) !== null && _b !== void 0 ? _b : aws_apigatewayv2_1.HttpMethod.ANY,
            payloadFormatVersion: aws_apigatewayv2_1.PayloadFormatVersion.VERSION_1_0,
            type: aws_apigatewayv2_1.HttpIntegrationType.HTTP_PROXY,
            uri: this.props.url,
            parameterMapping: this.props.parameterMapping,
        };
    }
}
exports.HttpProxyIntegration = HttpProxyIntegration;
_a = JSII_RTTI_SYMBOL_1;
HttpProxyIntegration[_a] = { fqn: "@aws-cdk/aws-apigatewayv2-integrations.HttpProxyIntegration", version: "1.134.0" };
